libra-node/src/main_node.rs

```diff
@@ -255,60 +255,68 @@ pub fn setup_environment(node_config: &mut NodeConfig) -> LibraHandle {
-                gen_simple_discovery_reconfig_subscription();
-            reconfig_subscriptions.push(simple_discovery_reconfig_subscription);
-            let network_config_listener =
-                ConfigurationChangeListener::new(conn_mgr_reqs_tx, RoleType::Validator);
-            runtime
-                .handle()
-                .spawn(network_config_listener.start(simple_discovery_reconfig_rx));
-        };
-
-        let (state_sync_sender, state_sync_events) =
-            state_synchronizer::network::add_to_network(&mut network_builder);
-        state_sync_network_handles.push((peer_id, state_sync_sender, state_sync_events));
-
-        let (mempool_sender, mempool_events) = libra_mempool::network::add_to_network(
-            &mut network_builder,
-            node_config.mempool.max_broadcasts_per_peer,
-        );
-        mempool_network_handles.push((peer_id, mempool_sender, mempool_events));
-        validator_network_provider = Some((peer_id, runtime, network_builder));
+    // Gather all network configs into a single vector.
+    // TODO:  consider explicitly encoding the role in the NetworkConfig
+    let mut network_configs: Vec<(RoleType, &mut NetworkConfig)> = node_config
+        .full_node_networks
+        .iter_mut()
+        .map(|network_config| (RoleType::FullNode, network_config))
+        .collect();
+    if let Some(network_config) = node_config.validator_network.as_mut() {
+        network_configs.push((RoleType::Validator, network_config));
     }
 
-    for mut full_node_network in node_config.full_node_networks.iter_mut() {
-        let (runtime, mut network_builder) = setup_network(
-            &mut full_node_network,
-            RoleType::FullNode,
-            Arc::clone(&db_rw.reader),
-            waypoint,
-        );
+    let mut validator_count = 0;
+    // Instantiate every network and collect the requisite endpoints for state_sync, mempool, and consensus.
+    for (role, network_config) in network_configs {
+        // Perform common instantiation steps
+        let (runtime, mut network_builder) =
+            setup_network(network_config, role, Arc::clone(&db_rw.reader), waypoint);
         let peer_id = network_builder.peer_id();
 
-        network_runtimes.push(runtime);
+        // Create the endpoints to connect the Network to StateSynchronizer.
         let (state_sync_sender, state_sync_events) =
             state_synchronizer::network::add_to_network(&mut network_builder);
         state_sync_network_handles.push((peer_id, state_sync_sender, state_sync_events));
+
+        // Create the endpoints to connect the network to MemPool.
         let (mempool_sender, mempool_events) = libra_mempool::network::add_to_network(
             &mut network_builder,
             node_config.mempool.max_broadcasts_per_peer,
         );
         mempool_network_handles.push((peer_id, mempool_sender, mempool_events));
 
```

Comment:
rather than merge these into a single vector of network_configs it might make sense to move this logic to setup_network.

Reply:
These should definitely stay outside of 'setup_network' (at least in its current form.

Taking a step back, I'm aiming to apply the following principles:

1. Make building hierarchical
1. main_node is the 'root' builder for the entire system.
1. move setup_network out of the root and into network
1. ensure that a builder explicitly references 'child' components but not 'sibling' components.   consensus and networking are siblings;  healthchecker is a child of networking.

Reply:
that's fair, forgive me for nitpicking here, but you know once you finish this, no one will revisit this code except to come back and complain it is too confusing -- without realizing how much more confusing it was :)

Reply:
A couple of outcomes come up:

1). I finish the work and the sequence of ~10 PRs needed to get there including documenting the final state.

2).  I don't finish the work and the next person tries to make it cleaner working from a moderately improved state all the while cursing my name.

Either way we're in a better position moving forward.

Reply:
well... that's an unfortunately reasonable argument.